src/com/amazon/ion/impl/ResizingPipedInputStream.java

```diff
@@ -0,0 +1,557 @@
+ * <ol>
+ *     <li>There is no risk of deadlock. Piped streams, which are intended for producing data in one
+ *     thread and consuming it in another, will block on read when no data is available and block on write
+ *     when the buffer is full. In a single-threaded context, avoiding deadlock on read requires checking
+ *     that bytes are available before every read. Avoiding deadlock on write would require checking that
+ *     the buffer is not full before every write, but there is no built-in, publicly-accessible way of doing
+ *     this with a PipedInputStream/PipedOutputStream.</li>
+ *     <li>The buffer can grow. Piped streams use a fixed-size buffer that causes blocking when full. If
+ *     used in a single-thread, this serves as a hard limit on the amount of data that can be written without
+ *     a matching read. This can require arbitrary limits on data size to be imposed by the application. The
+ *     NonBlockingCircularInputStream imposes no such limitation, but optionally allows for a maximum buffer
+ *     size to be configured to protect against unbounded growth.</li>
+ * </ol>
+ */
+public class ResizingPipedInputStream extends InputStream {
+
+    /**
+     * Handler of notifications provided by the ResizingPipedInputStream.
+     */
+    interface NotificationConsumer {
+
+        /**
+         * Bytes have been shifted to the start of the buffer in order to make room for additional bytes
+         * to be buffered.
+         * @param leftShiftAmount the amount of the left shift (also: the pre-shift read index of the first shifted
+         *                        byte).
+         */
+        void bytesConsolidatedToStartOfBuffer(int leftShiftAmount);
+    }
+
+    /**
+     * A NotificationConsumer that does nothing.
+     */
+    private static final NotificationConsumer NO_OP_NOTIFICATION_CONSUMER = new NotificationConsumer() {
+        @Override
+        public void bytesConsolidatedToStartOfBuffer(int leftShiftAmount) {
+            // Do nothing.
+        }
+    };
+
+    /**
+     * Mask to isolate a single byte.
+     */
+    private static final int SINGLE_BYTE_MASK = 0xFF;
+
+    /**
+     * The initial size of the buffer and the number of bytes by which the size of the buffer will increase
+     * each time it grows.
+     */
+    private final int initialBufferSize;
+
+    /**
+     * The maximum size of the buffer. If the user attempts to buffer more bytes than this, an exception will be raised.
+     */
+    private final int maximumBufferSize;
+
+    /**
+     * Whether to use a boundary to limit the number of available bytes. This can be used to buffer
+     * arbitrarily-sized chunks of bytes without making them available for consumption. When true,
+     * the boundary must be manually extended (see {@link #extendBoundary(int)} to make these bytes
+     * available. When false, all buffered bytes will be available to read.
+     */
+    private final boolean useBoundary;
```

Comment:
Based solely on a read through of this class, it's not obvious to me what the use case for this feature is. If there's a high-level example use case you could add to this comment, that would be helpful.

Reply:
The best example is `IonReaderLookaheadBuffer`, which fills the `ResizingPipedInputStream` in large chunks, but only wants to make the bytes available for consumption from `ResizingPipedInputStream` after reading the bytes to determine that they are part of the current top-level Ion value.

I don't really want to refer to `IonReaderLookaheadBuffer` (or even Ion) in `ResizingPipedInputStream` since this is a standalone class, so I hesitate to add this example in the comment. For now I'm actually content with the existing comment: "This can be used to buffer arbitrarily-sized chunks of bytes without making them available for consumption."

Reply:
@jobarr-amzn There isn't really a constraint that requires it to be implemented this way (that I can think of). A wrapper adds an extra layer (it would have to override most of the methods in this class to handle the boundary logic), but the boundary logic itself is pretty simple.